MGMT-16629: Use custom query params and headers for OS image downloads

paul-maidment opened the pull request:
## Description
This PR provides the ability to add custom query parameters and headers to every http request made while downloading OS images.

Two environment variables are to be provided `OS_IMAGES_HTTP_REQUEST_QUERY_PARAMS` and `OS_IMAGES_HTTP_REQUEST_HEADERS` These variables are JSON representations of `map[string]string` where the key is the parameter (or header) name and the value is the parameter (or header) value.

Once provided, these values will be used in the HTTP request headers and the HTTP query parameters of any request made during the download of OS images.


## Testing
This code includes tests to ensure that the functionality works as intended. A manual test will also be performed in podman to ensure the behavior is as expected


## Assignees
/cc @carbonin 
/cc @omertuc 

## Links
This is part of the Epic https://issues.redhat.com/browse/MGMT-15408


## Checklist

- [x] Title and description added to both, commit and PR
- [x] Relevant issues have been associated
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit tests (note that code changes require unit tests)

omertuc commented on main.go:
```diff
@@ -49,6 +49,22 @@ var Options struct {
 	// Will default to the value held in HTTPS_CA_FILE unless overridden
 	AssistedServiceApiTrustedCAFile string `envconfig:"ASSISTED_SERVICE_API_TRUSTED_CA_FILE" default:""`
+
+	// OSImageDownloadHeaders contains a JSON encoded representation of any HTTP headers to be sent with every request to download an image.
```
```suggestion
	// OSImageDownloadHeaders contains a JSON encoded representation of any
	// HTTP headers to be sent with every request to download an os image.

```

omertuc commented on main.go:
```diff
@@ -49,6 +49,22 @@ var Options struct {
+
+	// OSImageDownloadHeaders contains a JSON encoded representation of any HTTP headers to be sent with every request to download an image.
+	OSImageDownloadHeaders string `envconfig:"OS_IMAGES_HTTP_REQUEST_HEADERS" default:""`
+	// OSImageDownloadQueryParams contains a JSON encoded representation of any query parameters to be sent with every request to download an image.
```
```suggestion
	// OSImageDownloadQueryParams contains a JSON encoded representation of any
	// query parameters to be sent with every request to download an OS image.
```

omertuc commented on main.go:
```diff
@@ -49,6 +49,22 @@ var Options struct {
+func unmarshallJSONMap(jsonMap string) (map[string]string, error) {
+	result := make(map[string]string, 0)
+	if jsonMap != "" {
+		err := json.Unmarshal([]byte(jsonMap), &result)
+		if err != nil {
```
```suggestion
		if err := json.Unmarshal([]byte(jsonMap), &result); err != nil {
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -73,19 +73,21 @@ type ImageStore interface {
 )
 
-func NewImageStore(ed isoeditor.Editor, dataDir, imageServiceBaseURL string, insecureSkipVerify bool, versions []map[string]string, osImageDownloadTrustedCAFile string) (ImageStore, error) {
+func NewImageStore(ed isoeditor.Editor, dataDir, imageServiceBaseURL string, insecureSkipVerify bool, versions []map[string]string, osImageDownloadTrustedCAFile string, osImageDownloadHeadersMap map[string]string, osImageDownloadQueryParamsMap map[string]string) (ImageStore, error) {
```
```suggestion
func NewImageStore(ed isoeditor.Editor, dataDir, imageServiceBaseURL string, insecureSkipVerify bool, versions []map[string]string,
    osImageDownloadTrustedCAFile string, osImageDownloadHeadersMap map[string]string, osImageDownloadQueryParamsMap map[string]string) (ImageStore, error) {
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
 
-func (s *rhcosStore) downloadURLToFile(url string, path string) error {
-	resp, err := s.httpClient.Get(url)
+func (s *rhcosStore) getHttpResponse(url string) (*http.Response, error) {
```
I find `getHttpResponse` to be an odd name for something that makes an HTTP request, I think if we call it `doHttpRequest` or `performHTTPRequest` or something, the response return value is already implied

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
+	if err != nil {
+		return nil, err
+	}
+	// Apply any headers
```
Comment unnecessary
```suggestion
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
+	for key, value := range s.osImageDownloadHeadersMap {
+		req.Header.Set(key, value)
+	}
+	// Apply any query parameters
```
Comment unnecessary

```suggestion
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
+		}
+		req.URL.RawQuery = query.Encode()
+	}
+	// Send the query and obtain response
```
Comment unnecessary

```suggestion
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
+func (s *rhcosStore) getHttpResponse(url string) (*http.Response, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
```
Wrap the error

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,9 +151,33 @@ func validateVersions(versions []map[string]string) error {
+	// Send the query and obtain response
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return nil, err
```
Wrap the error

omertuc reviewed (changes requested):
Looks great apart for some cosmetic nits

omertuc commented on main.go:
```diff
@@ -49,6 +49,23 @@ var Options struct {
 	AssistedServiceApiTrustedCAFile string `envconfig:"ASSISTED_SERVICE_API_TRUSTED_CA_FILE" default:""`
+
+	// OSImageDownloadHeaders contains a JSON encoded representation of any
+	// HTTP headers to be sent with every request to download an os image.
```
```suggestion
	// HTTP headers to be sent with every request to download an OS image.
```

omertuc commented on pkg/imagestore/imagestore.go:
```diff
@@ -147,11 +153,32 @@ func validateVersions(versions []map[string]string) error {
+	resp, err := s.doHttpRequest(url)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "error encountered while making http request to %s, error is: %s", url, err.Error())
```
```suggestion
		return fmt.Errorf(err, "http request to %s failed: %w", url, err)
```

omertuc commented:
/lgtm

omertuc commented:
/approve

carbonin commented:
/hold

I have a few comments

carbonin commented on main.go:
```diff
@@ -49,6 +49,23 @@ var Options struct {
+
+	// OSImageDownloadHeaders contains a JSON encoded representation of any
+	// HTTP headers to be sent with every request to download an OS image.
+	OSImageDownloadHeaders string `envconfig:"OS_IMAGES_HTTP_REQUEST_HEADERS" default:""`
```
Maybe the env var name should match the variable name here?

paul-maidment replied:
Done

carbonin commented on main.go:
```diff
@@ -49,6 +49,23 @@ var Options struct {
+	OSImageDownloadQueryParams string `envconfig:"OS_IMAGES_HTTP_REQUEST_QUERY_PARAMS" default:""`
+}
+
+func unmarshallJSONMap(jsonMap string) (map[string]string, error) {
```
I thought envconfig handled maps? Did you look into that?

carbonin replied:
Ah, answered my own question. It doesn't quite use json  https://github.com/kelseyhightower/envconfig/blob/v1.4.0/envconfig.go#L319-L342

Probably best to stick with json then.
Also I suppose we're doing this already for OS Images :shrug:

paul-maidment replied:
Yeah, just changed it now to get rid of this

carbonin reviewed (commented):
Some rather minor comments.

If envconfig supports json maps I'd like to use that over deserializing manually

carbonin commented:
Okay given that the only comment is variable naming I'll leave it up to you @paul-maidment if you want to change that.
If not just remove the hold.

paul-maidment commented:
> Okay given that the only comment is variable naming I'll leave it up to you @paul-maidment if you want to change that. If not just remove the hold.

I renamed the variable and I also changed the way we bring the map in, so that we use envconfig directly for that.

carbonin commented:
I think the env var processing needs to go back to the way it was.
envconfig doesn't actually parse json for maps it seems

paul-maidment commented:
/test edge-e2e-metal-assisted-deploy-nodes

carbonin commented:
/unhold
